File: src/lowlevel/server/icefsoperators/icefsReadDirPlus.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 18:02:13
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsReadDirPlus.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
)

type IcefsReadDirPlusAppendEntryData func(fakeInode uint64, dir *IcefsDir, dirEntry any, dataSlice any) (status int32)

func (s *IcefsServer) GRpcIcefsReadDirPlusAppendEntryData(fakeInode uint64, dir *IcefsDir, dirEntry any, dataSlice any) (status int32) {
	var err error
	var entry any
	data := new(pb.IcefsReadDirPlusResDataPlus)
	data.DirEntry = dirEntry.(*pb.DirentStruct)

	dir.offset = data.DirEntry.Off

	if checkNameIsDotOrDotDot(data.DirEntry.Name) {
		return
	}

	entry, err = s.doIcefsLookUp(fakeInode, data.DirEntry.Name, GRpcFuseEntryParamBuilder)
	if err != nil {
		status = icefserror.IcefsStdErrno(err)
		return
	}

	data.Entry = entry.(*pb.FuseEntryParam)

	realDataSlice := dataSlice.(*[]*pb.IcefsReadDirPlusResDataPlus)

	*realDataSlice = append(*realDataSlice, data)
	status = icefserror.ICEFS_EOK
	return
}

func (s *IcefsServer) ThriftIcefsReadDirPlusAppendEntryData(fakeInode uint64, dir *IcefsDir, dirEntry any, dataSlice any) (status int32) {
	var err error
	var entry any
	data := new(icefsthrift.IcefsReadDirPlusData)
	data.DirEntry = dirEntry.(*icefsthrift.DirentStruct)

	dir.offset = data.DirEntry.Off

	if checkNameIsDotOrDotDot(data.DirEntry.Name) {
		return
	}

	entry, err = s.doIcefsLookUp(fakeInode, data.DirEntry.Name, ThriftFuseEntryParamBuilder)
	if err != nil {
		status = icefserror.IcefsStdErrno(err)
		return
	}

	data.Entry = entry.(*icefsthrift.FuseEntryParam)

	realDataSlice := dataSlice.(*[]*icefsthrift.IcefsReadDirPlusData)

	*realDataSlice = append((*realDataSlice), data)
	status = icefserror.ICEFS_EOK
	return
}

func (s *IcefsServer) doIcefsReadDirPlus(fakeInode uint64, offset int64, fh uint64, dataSlice any, dirStructBuilder DirStructBuilder, icefsReadDirPlusAppendEntryData IcefsReadDirPlusAppendEntryData) (status int32) {
	var dirEntry any
	var errno int32

	s.dirCacheLock.RLock()
	dir := s.getIcefsDir(fh)
	if dir == nil {
		s.dirCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}
	dir.dirLock.Lock()
	s.dirCacheLock.RUnlock()
	if offset != dir.offset {
		IcefsSeekDir(dir.dirStream, offset)
		dir.offset = offset
	}

	for {
		dirEntry, errno = IcefsReadDir(dir.dirStream, dirStructBuilder)
		if dirEntry == nil {
			if errno != icefserror.ICEFS_EOK {
				status = errno
				dir.dirLock.Unlock()
				return
			}
			break
		}

		status = icefsReadDirPlusAppendEntryData(fakeInode, dir, dirEntry, dataSlice)
		if status != icefserror.ICEFS_EOK {
			dir.dirLock.Unlock()
			return
		}
	}

	dir.dirLock.Unlock()
	status = icefserror.ICEFS_EOK
	return
}

func (s *IcefsGRpcServer) DoIcefsReadDirPlus(ctx context.Context, req *pb.IcefsReadDirPlusReq) (*pb.IcefsReadDirPlusRes, error) {
	var res pb.IcefsReadDirPlusRes
	res.Data = make([]*pb.IcefsReadDirPlusResDataPlus, 0)

	res.Status = s.server.doIcefsReadDirPlus(req.Inode, req.Offset, req.Fh, &res.Data, GRpcDirStructBuilder, s.server.GRpcIcefsReadDirPlusAppendEntryData)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsReadDirPlus(ctx context.Context, req *icefsthrift.IcefsReadDirPlusReq) (*icefsthrift.IcefsReadDirPlusRes, error) {
	var res icefsthrift.IcefsReadDirPlusRes
	res.Data = make([]*icefsthrift.IcefsReadDirPlusData, 0)

	res.Status = s.server.doIcefsReadDirPlus(uint64(req.Inode), req.Offset, uint64(req.Fh), &res.Data, ThriftDirStructBuilder, s.server.ThriftIcefsReadDirPlusAppendEntryData)

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsReadDirPlus.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-04 15:53:44
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsReadDirPlus.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsReadDirPlus(fuse_req_t fuseReq, fuse_ino_t inode,
                                     size_t size, off_t offset,
                                     struct fuse_file_info *fi) {
  IcefsReadDirPlusReq req;
  IcefsReadDirPlusRes res;
  grpc::ClientContext ctx;
  ICEFS_PR_FUNCTION;
  req.set_inode(inode);
  req.set_offset(offset);
  req.set_fh(fi->fh);

  grpc::Status status = stub_->DoIcefsReadDirPlus(&ctx, req, &res);
  if (status.ok() && !res.status()) {
    size_t remain = size;
    char *buf = new (std::nothrow) char[size];
    char *p = buf;
    size_t entrySize;
    int dataSize = res.data_size();
    struct fuse_entry_param entry = {0};
    for (int i = 0; i < dataSize; ++i) {
      IcefsFillFuseEntryParamIn(&entry, res.data(i).entry());
      entrySize = fuse_add_direntry_plus(fuseReq, p, remain,
                                         res.data(i).dir_entry().name().c_str(),
                                         &entry, res.data(i).dir_entry().off());
      if (entrySize > remain) {
        // TODO: forget one
        break;
      }
      p += entrySize;
      remain -= entrySize;
    }
    int32_t ret = fuse_reply_buf(fuseReq, buf, size - remain);
    if (ret != ICEFS_EOK) {
      fuse_reply_err(fuseReq, ret);
      std::cout << "ERR:" << __FUNCTION__ << "fuse_reply_buf" << ret
                << std::endl;
    }
    delete[] buf;
  } else {
    fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
    ICEFS_PR_ERR_STATUS;
  }
}